### PR TITLE
24 encryption

### PR DESCRIPTION
**Pull request by dwalintukan**
1. renamed `Login` -> `AccountLogin`
2. `Login` is now the very first screen. the login with pw/verification.

image
image
image
image

### CODE REVIEW

File: src/stores/WalletStore.ts [hunk 1]
```diff
@@ -19,42 +24,76 @@ export default class WalletStore {
...
+    const passwordHash = derivedKey.toString('hex');
+
+    if (!this.passwordHash) {
+      // New user, set passwordHash in storage
+      this.passwordHash = passwordHash;
+      chrome.storage.local.set(
+        { [STORAGE.PASSWORD_HASH]: passwordHash },
+        () => console.log('passwordHash set'),
+      );
+      this.routeToAccountPage();
```
Comment (hayeah) [inline src/stores/WalletStore.ts]: return, and we can remove an "else" nesting.

File: src/stores/WalletStore.ts [hunk 2]
```diff
@@ -19,42 +24,76 @@ export default class WalletStore {
...
+  * Creates a new passwordHash or validates the existing one for the main login.
+  * @param password {string} The new/existing password for the per-install appSalt.
+  */
+  @action
+  public login = async (password: string) => {
+    this.loading = true;
+
+    // TODO: make this async so it doesnt block the UI
+    // Generate appSalt if needed
+    if (!this.appSalt) {
```
Comment (hayeah) [inline src/stores/WalletStore.ts]: perhaps abstract this into a method:

```
generateAndSetPasswordHash
```

File: src/stores/WalletStore.ts [hunk 3]
```diff
@@ -19,42 +24,76 @@ export default class WalletStore {
...
+      const appSalt: Uint8Array = window.crypto.getRandomValues(new Uint8Array(16)) as Uint8Array;
+      this.appSalt = appSalt;
+      chrome.storage.local.set(
+        { [STORAGE.APP_SALT]: appSalt.toString() },
+        () => console.log('appSalt set'),
+      );
+    }
+
+    // Derive passwordHash
+    const saltBuffer = Buffer.from(this.appSalt!);
```
Comment (hayeah) [inline src/stores/WalletStore.ts]: maybe create a getter for `passwordHash`?

Comment (hayeah) [reply src/stores/WalletStore.ts]: password hash should throw error if appSalt is empty

File: src/stores/WalletStore.ts [hunk 4]
```diff
@@ -81,75 +120,123 @@ export default class WalletStore {
...
+  private fetchStorageValues = () => {
+    const { APP_SALT, PASSWORD_HASH, MAINNET_ACCOUNTS, TESTNET_ACCOUNTS } = STORAGE;
+    chrome.storage.local.get([APP_SALT, PASSWORD_HASH, MAINNET_ACCOUNTS, TESTNET_ACCOUNTS],
+      ({ appSalt, passwordHash, mainnetAccounts, testnetAccounts }: any) => {
+        if (!isEmpty(appSalt)) {
+          const array = split(appSalt, ',').map((str) => parseInt(str, 10));
+          this.appSalt =  Uint8Array.from(array);
+        }
+
+        if (!isEmpty(passwordHash)) {
```
Comment (hayeah) [inline src/stores/WalletStore.ts]: hmmm, i think we shouldn't store passwordHash in storage? It should only be kept in memory.

File: src/stores/WalletStore.ts (hunk 1 above)
Comment (dwalintukan) [reply src/stores/WalletStore.ts]: done

File: src/stores/WalletStore.ts (hunk 2 above)
Comment (dwalintukan) [reply src/stores/WalletStore.ts]: done

File: src/stores/WalletStore.ts (hunk 3 above)
Comment (dwalintukan) [reply src/stores/WalletStore.ts]: done

File: src/stores/WalletStore.ts (hunk 4 above)
Comment (dwalintukan) [reply src/stores/WalletStore.ts]: as discussed, will remove this and change the flow for the first time user

Comment (dwalintukan) [reply src/stores/WalletStore.ts]: I ended up keeping the `Login` page as the initial page with a note that the master pw wont save unless they create/import their first account.

File: src/stores/WalletStore.ts [hunk 5]
```diff
@@ -61,40 +73,29 @@ export default class WalletStore {
...
-    // Generate appSalt if needed
-    if (!this.appSalt) {
-      const appSalt: Uint8Array = window.crypto.getRandomValues(new Uint8Array(16)) as Uint8Array;
-      this.appSalt = appSalt;
-      chrome.storage.local.set(
-        { [STORAGE.APP_SALT]: appSalt.toString() },
-        () => console.log('appSalt set'),
-      );
+    // TODO: find solution to unblock UI
+    this.generateAppSalt();
```
Comment (hayeah) [inline src/stores/WalletStore.ts]: nitpick. `generate` seems to indicate a new salt each time.

`this.getOrGenerateAppSalt();`

or

`this.getAppSalt()`

**Comment by hayeah** [comment]
lgtm

**Pull request merged**